// ripple-binary-codec/distrib/npm/index.js
'use strict';

var assert = require('assert');
var coreTypes = require('./coretypes');
var quality = coreTypes.quality;
var _coreTypes$binary = coreTypes.binary;
var bytesToHex = _coreTypes$binary.bytesToHex;
var signingData = _coreTypes$binary.signingData;
var multiSigningData = _coreTypes$binary.multiSigningData;
var binaryToJSON = _coreTypes$binary.binaryToJSON;
var serializeObject = _coreTypes$binary.serializeObject;

function decode(binary) {
  assert(typeof binary === 'string', 'binary must be a hex string');
  return binaryToJSON(binary);
}

function encode(json) {
  assert(typeof json === 'object');
  return bytesToHex(serializeObject(json));
}

function encodeForSigning(json) {
  assert(typeof json === 'object');
  return bytesToHex(signingData(json));
}

function encodeForMultisigning(json, signer) {
  assert(typeof json === 'object');
  assert.equal(json.SigningPubKey, '');
  return bytesToHex(multiSigningData(json, signer));
}

function encodeQuality(value) {
  assert(typeof value === 'string');
  return bytesToHex(quality.encode(value));
}

function decodeQuality(value) {
  assert(typeof value === 'string');
  return quality.decode(value).toString();
}

module.exports = {
  decode: decode,
  encode: encode,
  encodeForSigning: encodeForSigning,
  encodeForMultisigning: encodeForMultisigning,
  encodeQuality: encodeQuality,
  decodeQuality: decodeQuality
};
